File: ios/Bridge/Plugins/BarcodeScanningPlugin.swift
import Foundation

#if MLKIT_BARCODE_SCANNING
  #if canImport(VisionCamera)
    import VisionCamera

    @objc(BarcodeScanningPlugin)
    public class BarcodeScanningPlugin: BaseMLKitPlugin {

      private lazy var recognitionOptions: BarcodeScanningOptions = {
        BarcodeScanningOptions(
          formats: BarcodeScanningOptionParser.parseFormats(self.options?["formats"]),
          enableAllPotentialBarcodes: BarcodeScanningOptionParser.parseEnableAllPotentialBarcodes(
            self.options?["enableAllPotentialBarcodes"]
          ),
          invertColors: self.invertColors,
          scaleFactor: self.scaleFactor
        )
      }()

      private lazy var recognizeBarcodesUseCase: RecognizeBarcodesUseCase = {
        let recognitionService = BarcodeScanningServiceFactory.create(
          options: self.recognitionOptions)
        let imagePreprocessor = ImagePreprocessor()

        return RecognizeBarcodesUseCase(
          imagePreprocessor: imagePreprocessor,
          recognitionService: recognitionService
        )
      }()

      public override func processFrame(
        _ frame: Frame,
        withArguments arguments: [AnyHashable: Any]?
      ) throws -> [String: Any] {
        let options = BarcodeScanningOptions(
          formats: recognitionOptions.formats,
          enableAllPotentialBarcodes: recognitionOptions.enableAllPotentialBarcodes,
          invertColors: recognitionOptions.invertColors,
          outputOrientation: outputOrientation,
          scaleFactor: recognitionOptions.scaleFactor
        )

        _ = arguments
        let result = try recognizeBarcodesUseCase.execute(frame: frame, options: options)
        return BarcodeScanningSerializer.toReactNativeMap(result)
      }
    }
  #endif
#endif  // MLKIT_BARCODE_SCANNING

File: android/src/main/java/com/visioncameramlkit/bridge/plugins/BarcodeScanningPlugin.kt
package com.visioncameramlkit.bridge.plugins

import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.VisionCameraProxy
import com.visioncameramlkit.application.usecases.RecognizeBarcodesUseCase
import com.visioncameramlkit.bridge.parsers.BarcodeScanningOptionParser
import com.visioncameramlkit.domain.models.BarcodeScanningOptions
import com.visioncameramlkit.infrastructure.image.ImagePreprocessor
import com.visioncameramlkit.infrastructure.mlkit.factories.BarcodeScanningServiceFactory
import com.visioncameramlkit.infrastructure.serializers.BarcodeScanningSerializer

@Suppress("unused")
class BarcodeScanningPlugin(
  proxy: VisionCameraProxy,
  options: Map<String, Any>?,
) : BaseMLKitPlugin(proxy, options) {
  override val tag = "BarcodeScanningPlugin"

  private val barcodeScanningOptions: BarcodeScanningOptions =
    BarcodeScanningOptions(
      formats = BarcodeScanningOptionParser.parseFormats(options?.get("formats")),
      enableAllPotentialBarcodes =
        BarcodeScanningOptionParser.parseEnableAllPotentialBarcodes(
          options?.get("enableAllPotentialBarcodes"),
        ),
      invertColors = invertColors,
      outputOrientation = outputOrientation,
      scaleFactor = scaleFactor,
    )
  private val recognizeBarcodesUseCase: RecognizeBarcodesUseCase

  init {

    val recognitionService = BarcodeScanningServiceFactory.create(barcodeScanningOptions)
    val imagePreprocessor = ImagePreprocessor()

    recognizeBarcodesUseCase = RecognizeBarcodesUseCase(imagePreprocessor, recognitionService)
  }

  override fun processFrame(
    frame: Frame,
    arguments: Map<String, Any>?,
  ): HashMap<String, Any> {
    val result = recognizeBarcodesUseCase.execute(frame, barcodeScanningOptions)
    return BarcodeScanningSerializer.toReactNativeMap(result)
  }
}
